File: 2 Semestre (Estrutura de Dados)/Aula_3/Exercicio/mercado.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "mercado.h"

static int contador_produto = 0;
static int contador_carrinho = 0;
static Carrinho carrinho[50];
static Produto produtos[50];

void infoProduto(Produto prod) {
    printf("Cód: %d \nNome: %s \nPreço: %.2f\n", prod.codigo, prod.nome, prod.preco);
}

void menu() {
    printf("================ Bem-vindo(a) - Solução Mercado ===========\n");
    printf("Selecione uma opção abaixo: \n");
    printf("1 - Cadastrar produto\n");
    printf("2 - Listar produtos\n");
    printf("3 - Comprar produto\n");
    printf("4 - Visualizar carrinho\n");
    printf("5 - Fechar pedido\n");
    printf("6 - Sair do sistema\n");

    int opcao;
    scanf("%d", &opcao);
    getchar();

    switch (opcao) {
        case 1:
            cadastrarProduto();
            break;
        case 2:
            listarProdutos();
            break;
        case 3:
            comprarProduto();
            break;
        case 4:
            visualizarCarrinho();
            break;
        case 5:
            fecharPedido();
            break;
        case 6:
            printf("Volte sempre!\n");
            sleep(2);
            exit(0);
        default:
            printf("Opção inválida.\n");
            sleep(2);
            menu();
            break;
    }
}

void cadastrarProduto() {
    printf("Cadastro de Produto\n");
    printf("====================\n");

    printf("Informe o nome do produto: \n");
    fgets(produtos[contador_produto].nome, 30, stdin);
    produtos[contador_produto].nome[strcspn(produtos[contador_produto].nome, "\n")] = 0;

    printf("Informe o preço do produto: \n");
    scanf("%f", &produtos[contador_produto].preco);

    printf("O produto %s foi cadastrado com sucesso.\n", produtos[contador_produto].nome);
    produtos[contador_produto].codigo = (contador_produto + 1);
    contador_produto++;
    sleep(2);
    menu();
}

void listarProdutos() {
    if(contador_produto > 0) {
        printf("Listagem de produtos.\n");
        printf("---------------------\n");
        for(int i = 0; i < contador_produto; i++) {
            infoProduto(produtos[i]);
            printf("------------------\n");
            sleep(1);
        }
        sleep(2);
        menu();
    } else {
        printf("Não temos ainda produtos cadastrados.\n");
        sleep(2);
        menu();
    }
}

void comprarProduto() {
    if(contador_produto > 0) {
        printf("Informe o código do produto que deseja adicionar ao carrinho.\n");

        printf("========== Produtos Disponíveis ===========\n");
        for(int i = 0; i < contador_produto; i++) {
            infoProduto(produtos[i]);
            printf("---------------------\n");
            sleep(1);
        }
        int codigo;
        scanf("%d", &codigo);
        getchar();

        int tem_mercado = 0;
        for(int i = 0; i < contador_produto; i++) {
            if(produtos[i].codigo == codigo) {
                tem_mercado = 1;

                if(contador_carrinho > 0) {
                    int * retorno = temNoCarrinho(codigo);

                    if(retorno[0] == 1) {
                        carrinho[retorno[1]].quantidade++;
                        printf("Aumentei a quantidade do produto %s já existente no carrinho.\n",
                                strtok(carrinho[retorno[1]].produto.nome, "\n"));
                        sleep(2);
                        menu();
                    } else {
                        Produto p = pegarProdutoPorCodigo(codigo);
                        carrinho[contador_carrinho].produto = p;
                        carrinho[contador_carrinho].quantidade = 1;
                        contador_carrinho++;
                        printf("O produto %s foi adicionado ao carrinho.\n", strtok(p.nome, "\n"));
                        sleep(2);
                        menu();
                    }
                } else {
                    Produto p = pegarProdutoPorCodigo(codigo);
                    carrinho[contador_carrinho].produto = p;
                    carrinho[contador_carrinho].quantidade = 1;
                    contador_carrinho++;
                    printf("O produto %s foi adicionado ao carrinho.\n", strtok(p.nome, "\n"));
                    sleep(2);
                    menu();
                }
            }
        }
        if(tem_mercado < 1) {
            printf("Não foi encontrado o produto com código %d.\n", codigo);
            sleep(2);
            menu();
        }
    } else {
        printf("Ainda não existem produtos para vender.\n");
        sleep(2);
        menu();
    }
}

void visualizarCarrinho() {
    if(contador_carrinho > 0) {
        printf("Produtos do Carrinho\n");
        printf("--------------------\n");
        for(int i = 0; i < contador_carrinho; i++) {
            infoProduto(carrinho[i].produto);
            printf("Quantidade: %d\n", carrinho[i].quantidade);
            printf("-----------------\n");
            sleep(1);
        }
        sleep(2);
        menu();
    } else {
        printf("Não temos ainda produtos no carrinho.\n");
        sleep(2);
        menu();
    }
}

Produto pegarProdutoPorCodigo(int codigo) {
    Produto p;
    for(int i = 0; i < contador_produto; i++) {
        if(produtos[i].codigo == codigo) {
            p = produtos[i];
        }
    }
    return p;
}

int * temNoCarrinho(int codigo) {
    static int retorno[] = {0, 0};
    for(int i = 0; i < contador_carrinho; i++) {
        if(carrinho[i].produto.codigo == codigo) {
            retorno[0] = 1; //tem o produto com este código no carrinho
            retorno[1] = i; //o índice do produto no carrinho
        }
    }
    return retorno;
}

void fecharPedido() {
    if(contador_carrinho > 0) {
        float valorTotal = 0.0;
        printf("Produtos do Carrinho\n");
        printf("--------------------\n");
        for(int i = 0; i < contador_carrinho; i++) {
            Produto p = carrinho[i].produto;
            int quantidade = carrinho[i].quantidade;
            valorTotal += p.preco * quantidade;
            infoProduto(p);
            printf("Quantidade: %d\n", quantidade);
            printf("---------------\n");
            sleep(1);
        }
        printf("Sua fatura é R$ %.2f\n", valorTotal);

        //limpar carrinho
        contador_carrinho = 0;
        printf("Obrigado pela preferência.\n");
        sleep(5);
        menu();
    } else {
        printf("Você não tem nenhum produto no carrinho ainda.\n");
        sleep(3);
        menu();
    }
}

File: 1 Semestre/1 Semestre (C)/EXTRAS/mercado.py
class Produto:
    def __init__(self, codigo, nome, preco):
        self.codigo = codigo
        self.nome = nome
        self.preco = preco


class ItemCarrinho:
    def __init__(self, produto, quantidade=1):
        self.produto = produto
        self.quantidade = quantidade


def menu():
    print("\nMenu")
    print("1. Cadastrar Produto")
    print("2. Listar Produtos")
    print("3. Comprar Produto")
    print("4. Visualizar Carrinho")
    print("5. Fechar Pedido")
    print("6. Atualizar Produto")
    print("7. Sair do Sistema")


def cadastrar_produto(produtos):
    while True:
        print(f"\nCadastrar produto {len(produtos) + 1}")
        try:
            codigo = int(input("Digite o código do produto: "))
            nome = input("Digite o nome do produto: ")
            preco = float(input("Digite o preço do produto: "))
            produtos.append(Produto(codigo, nome, preco))
        except ValueError:
            print("Entrada inválida. Tente novamente.")
            continue

        continuar = input("Deseja cadastrar outro produto? (s/n): ").lower()
        if continuar == 'n':
            break


def listar_produtos(produtos):
    if not produtos:
        print("Nenhum produto cadastrado.")
        return

    print("\nProdutos disponíveis:")
    for p in produtos:
        print(f"{p.codigo} | {p.nome} | R$ {p.preco:.2f}")


def pegar_produto_por_codigo(produtos, codigo):
    for i, p in enumerate(produtos):
        if p.codigo == codigo:
            return i
    return -1


def tem_no_carrinho(carrinho, codigo):
    for i, item in enumerate(carrinho):
        if item.produto.codigo == codigo:
            return i
    return -1


def comprar_produto(produtos, carrinho):
    try:
        codigo = int(input("\nDigite o código do produto para comprar: "))
    except ValueError:
        print("Código inválido.")
        return

    pos = pegar_produto_por_codigo(produtos, codigo)
    if pos == -1:
        print("Produto não encontrado.")
        return

    pos_carrinho = tem_no_carrinho(carrinho, codigo)
    if pos_carrinho != -1:
        carrinho[pos_carrinho].quantidade += 1
    else:
        print("\nProduto adicionado ao carrinho.")
        carrinho.append(ItemCarrinho(produtos[pos]))


def visualizar_carrinho(carrinho):
    if not carrinho:
        print("\nCarrinho vazio.")
        return

    print("\nItens no carrinho:")
    for item in carrinho:
        print(f"\nCódigo: {item.produto.codigo} | Nome: {item.produto.nome} | "
              f"Preço: R$ {item.produto.preco:.2f} | Quantidade: {item.quantidade}")


def fechar_pedido(carrinho):
    if not carrinho:
        print("\nCarrinho vazio.")
        return

    total = 0.0
    print("\nValor do Pedido:")
    for item in carrinho:
        subtotal = item.produto.preco * item.quantidade
        print(f"Produto: {item.produto.nome} | Preço: R$ {item.produto.preco:.2f} "
              f"| Quantidade: {item.quantidade} | Total: R$ {subtotal:.2f}")
        total += subtotal

    print(f"\nTotal da compra: R$ {total:.2f}")
    print("Pedido fechado com sucesso! Carrinho esvaziado.")
    carrinho.clear()


def atualizar_produto(produtos, carrinho):
    try:
        codigo = int(input("\nDigite o código do produto que deseja atualizar: "))
    except ValueError:
        print("Código inválido.")
        return

    pos = pegar_produto_por_codigo(produtos, codigo)
    if pos == -1:
        print("Produto não encontrado.")
        return

    print(f"\nProduto encontrado: {produtos[pos].nome} | R$ {produtos[pos].preco:.2f}")
    novo_nome = input("Digite o novo nome do produto: ")
    try:
        novo_preco = float(input("Digite o novo preço do produto: "))
    except ValueError:
        print("Preço inválido.")
        return

    produtos[pos].nome = novo_nome
    produtos[pos].preco = novo_preco
    print("Produto atualizado com sucesso!")

    for item in carrinho:
        if item.produto.codigo == codigo:
            item.produto = produtos[pos]


def main():
    produtos = []
    carrinho = []

    while True:
        menu()
        try:
            opcao = int(input("Escolha uma opção: "))
        except ValueError:
            print("Entrada inválida. Tente novamente.")
            continue

        if opcao == 1:
            cadastrar_produto(produtos)
        elif opcao == 2:
            listar_produtos(produtos)
        elif opcao == 3:
            comprar_produto(produtos, carrinho)
        elif opcao == 4:
            visualizar_carrinho(carrinho)
        elif opcao == 5:
            fechar_pedido(carrinho)
        elif opcao == 6:
            atualizar_produto(produtos, carrinho)
        elif opcao == 7:
            print("O sistema está sendo encerrado!")
            break
        else:
            print("Opção inválida! Por favor, tente novamente.")


if __name__ == "__main__":
    main()
